src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+        goto error;
+    result = set_current_trigger(device, trigger_name);
+
+    if (result)
+        device->trigger_name = strdup(trigger_name);
+
+    free(trigger_name);
+
+    return result;
+
+error:
+    SOL_WRN("Could not create sysfs trigger.");
+
+    return false;
+}
+
+static bool
+check_trigger(struct sol_iio_device *device)
+{
+    char path[PATH_MAX + 1];
+    char *name = NULL;
+    int len;
+
+    len = snprintf(path, sizeof(path), CURRENT_TRIGGER_DEVICE_PATH, device->device_id);
+    if (len < 0 || len > PATH_MAX) {
+        SOL_WRN("Could not open IIO device%d trigger on sysfs", device->device_id);
+        return false;
+    }
+
+    len = sol_util_read_file(path, "%ms", &name);
+    if (len < 0) {
+        SOL_INF("No current trigger for iio:device%d. Creating a sysfs one.",
+            device->device_id);
+        if (!create_sysfs_trigger(device))
+            return false;
+    } else
+        device->trigger_name = name;
+
+    return check_manual_triggering(device);
+}
+
+static void
+set_buffer_size(struct sol_iio_device *device, int buffer_size)
+{
+    char path[PATH_MAX + 1];
+
+    if (craft_filename_path(path, sizeof(path), BUFFER_LENGHT_DEVICE_PATH, device->device_id)) {
+        SOL_WRN("Could not set IIO device buffer size");
+        return;
+    }
+
+    if (!write_file(path, "%d", buffer_size)) {
+        SOL_WRN("Could not set IIO device buffer size");
+    }
+
+    return;
+}
+
+static bool
+set_buffer_enabled(struct sol_iio_device *device, bool enabled)
+{
+    char path[PATH_MAX + 1];
+    int len;
+
+    len = snprintf(path, sizeof(path), BUFFER_ENABLE_DEVICE_PATH, device->device_id);
+    if (len < 0 || len > PATH_MAX)
+        return false;
+
+    if (!write_file(path, "%d", enabled))
+        return false;
+
+    return true;
+}
+
+static size_t
+calc_buffer_size(struct sol_iio_device *device)
+{
+    struct sol_iio_channel *channel;
+    int i;
+    size_t size = 0;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&device->channels, channel, i) {
+        size += channel->storagebits;
+    }
+
+    return size / 8;
+}
+
+static bool
+device_reader_cb(void *data, int fd, unsigned int active_flags)
+{
+    struct sol_iio_device *device = data;
+    bool result = true;
+    size_t len = calc_buffer_size(device);
+    uint8_t *buffer = NULL;
+
+    buffer = calloc(1, len);
+    if (!buffer) {
+        SOL_WRN("Could not allocate memory for read buffer");
+        result = false;
+    }
+
+    if (result && active_flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP | SOL_FD_FLAGS_NVAL)) {
+        SOL_WRN("Unexpected reading");
+        result = false;
+    }
+
+    while (result) {
+        ssize_t ret;
+
+        ret = read(fd, buffer, len);
+        if (ret < 0) {
+            if (errno == EINTR)
+                continue;
+            else if (errno == EAGAIN || errno == EWOULDBLOCK)
+                break;
+            else {
+                result = false;
+                break;
+            }
+        }
```

Comment:
Sounds good. I'm thinking, however, of doing this as a separate PR, after this one goes in, as it will touch different places. Is it ok?